Fix Camera.side() calculation and improve Camera

HannesWell opened the pull request:
With this PR I want to submit a fix for the camera side calculation. The scale of the data was not considered and the side calculation itself was (for me) not completely comprehensible. I think there was the assumption that the up-direction is always (0,0,1).

My point about the side calculation can be seen easily with the `org.jzy3d.demos.scatter.ScatterDemo` if you use the following `init()` method with that fixed points:
>     public void init() {
> 
>         Coord3d[] points = new Coord3d[8];
> 
>         points[0] = new Coord3d(7.0404835, 905.0, 53.014378);
>         points[1] = new Coord3d(7.0043426, 935.0, 53.014378);
>         points[2] = new Coord3d(7.0404835, 905.0, 53.014378);
>         points[3] = new Coord3d(6.9886136, 885.0, 151.9809);
>         points[4] = new Coord3d(7.0043426, 935.0, 53.014378);
>         points[5] = new Coord3d(6.96367, 915.0, 153.10077);
>         points[6] = new Coord3d(6.9886136, 885.0, 151.9809);
>         points[7] = new Coord3d(6.96367, 915.0, 153.10077);
> 
>         Scatter scatter = new Scatter(points, Color.BLUE, 5.0f);
>         chart = AWTChartComponentFactory.chart(Quality.Advanced, "newt");
>         chart.getScene().add(scatter);
>     }
> 
The z-axis label should be at the right side, but jumps from right to left and back when you rotate the diagram (mainly around the z-axis).

To fix this issue I added a `scale` field to `Camera` and additionally improved the structure how the camera position is set (all parameters are set together with `setPosition()`) and made the old methods deprecated (If you want to keep them, I can remove the deprecation). Additionally I improved the structure of the calling methods in Axis and split `computeCameraUpAndTriggerEvents()` in two methods, without changing the behaviour.


In the second commit I add three flags to `AxeBox` to control if axis-labels and -ticks are drawn at the bottom or top respectively left or right. I personally like the z-labels and -ticks on the left. With that commit the user can choose his preference for that. The default is the previous behaviour (labels/ticks at the bottom respectively right). 
If you like it, I can make this part of the `IAxe` interface or maybe better of the `IAxeLayout` interface.

HannesWell commented:
Do you have any remarks on this PR?

jzy3d commented:
All my apologies for the very late consideration of your work. This is a great addition to the library.

To answer you : yes cam.side() assume up vector is always (0,0,1)

Here are my suggestions after reading your changes.
- the camera object has an existing "knowledge" of the view scale that can be seen in the Camera.getDistance() method. This is especially usefull for processing distance of polygons to the viewpoint, in order to choose the rendering order in the case of charts with alpha. We should probably only use the scale value held by the camera instead of receiving this value from the caller of Camera.getDistance().  
- I would move the booleans holding the axis layout configuration (left/right, up/down position of ticks) to the AxeLayout class since it already store all layout settings of an axis. Being able to change this is cool.
- I need to make tests to ensure this will work properly with the logarithmic library that uses spaceTransform and apply changes to the axis center processing and ensure this remain compatible (I think it should be OK).

I would love to see few unit tests on this. I know that sound weird since the library has almost no test. I am currently making big changes to the lib and try to add more integration and unit tests with mocks. Despite having created a test tool that allows comparing chart images with a reference, I am  now trying to rely more on unit tests that are more "stable".

HannesWell commented:
No problem. I'm glad you like it.

I'll apply your remarks as soon as possible, but unfortunately in the next two or three weeks I probably won't have the time for it. And it has been a while I looked into the code.^^

Regarding the test, I'll try to write some good one. I was wondering how you test the library. It's great to have good tests but I know the pain it can be to write them. But if you don't have them, sooner or later it's a pain too.

jzy3d commented:
I am currently upgrading the framework (a surprise is coming :) I am adding at the same time some tests with an ability to mock part of the chart. Maybe this will be helpful to you. This work is not publish yet. I will send you example tests as soon as the work is published.

HannesWell commented:
OK, great. I'm curious about it.

jzy3d commented:
Hi,
I tried your scatter you submitter on current master branch and could not reproduce the issue.

It displays as follow. No matter how I shake the axis box, the Z tick labels and axis label stick on the right side as expected.

![image](https://user-images.githubusercontent.com/2334960/102548902-9bb00c00-40bb-11eb-9401-282140d491e7.png)

Did I misunderstood your problem?

I run the following code

```
public class ScatterDemo extends AbstractAnalysis{
	public static void main(String[] args) throws Exception {
		AnalysisLauncher.open(new ScatterDemo());
	}
		
	@Override
    public void init(){
        Coord3d[] points = new Coord3d[8];

        points[0] = new Coord3d(7.0404835, 905.0, 53.014378);
        points[1] = new Coord3d(7.0043426, 935.0, 53.014378);
        points[2] = new Coord3d(7.0404835, 905.0, 53.014378);
        points[3] = new Coord3d(6.9886136, 885.0, 151.9809);
        points[4] = new Coord3d(7.0043426, 935.0, 53.014378);
        points[5] = new Coord3d(6.96367, 915.0, 153.10077);
        points[6] = new Coord3d(6.9886136, 885.0, 151.9809);
        points[7] = new Coord3d(6.96367, 915.0, 153.10077);

        Scatter scatter = new Scatter(points, Color.BLUE, 5.0f);
        chart = AWTChartComponentFactory.chart(Quality.Advanced, "newt");
        chart.getScene().add(scatter);
    }
}

```

HannesWell commented:
Now I have to apologize for my later answer.

Yes you understand my problem right. But it's strange. In the same point of view the z-axis label is displayed on the left:

![grafik](https://user-images.githubusercontent.com/44067969/105768865-4ccc8080-5f5d-11eb-803b-928a63155a47.png)

Used the exactly the code you posted with the release commit of version 1.0.3 (edfb6c521b95d50e866a61c941a0e35cb1f34441) and the result is like in my picture. If I rotate the view around the z-axis (graph in the middle of the box and drag to the left or right) the jumps to the right after a certain rotation angle, then later it jumps back.
I can capture a video of it.

I also verified it for the current master (commit b0564e8abaa926b5399d6746ec78c022588b5fb6) and the behaviour is still the same for me. I used the following code in the `ScatterDemoAWT`:
```
public class ScatterDemoAWT extends AWTAbstractAnalysis {
	public static void main(String[] args) throws Exception {
		AnalysisLauncher.open(new ScatterDemoAWT());
	}

	public void init() {
		Coord3d[] points = new Coord3d[8];

		points[0] = new Coord3d(7.0404835, 905.0, 53.014378);
		points[1] = new Coord3d(7.0043426, 935.0, 53.014378);
		points[2] = new Coord3d(7.0404835, 905.0, 53.014378);
		points[3] = new Coord3d(6.9886136, 885.0, 151.9809);
		points[4] = new Coord3d(7.0043426, 935.0, 53.014378);
		points[5] = new Coord3d(6.96367, 915.0, 153.10077);
		points[6] = new Coord3d(6.9886136, 885.0, 151.9809);
		points[7] = new Coord3d(6.96367, 915.0, 153.10077);

		Scatter scatter = new Scatter(points, Color.BLUE, 5.0f);
		chart = AWTChartFactory.chart(Quality.Advanced);
		chart.getScene().add(scatter);
	}
}
```

I don't know why it behaves different on my machine.
I'm using Windows10 and the GraalVM 20.3.0, but I think this is isn't a platform related issue because with the proposed changes the issue was fixed for me.

But I will rebase the proposed changes to the current master in the next days and investigate it further.

jzy3d commented:
Hi, just to let you know, I have added in version 2.0 something to allow choosing between Left or Right side for Z label.

https://github.com/jzy3d/jzy3d-api/commit/a8eab7856bc35b263a9f5c90ae9e4efd4353844e

Does this fix your problem?

HannesWell commented:
Hi thanks for this improvement.
I've updated this PR and removed the commit for the axis labels.
Unfortunately the Camera.side() method is still broken and the z-axis label still jumps between left and right.
I've updated my fix for the `Camera.side` computation to the current master branch.

I still have no idea why this doesn't work for me, but works for you (maybe the example is to fragile and something on your computer is different that makes it work).

When I run the the following code:
```
public class ScatterDemoAWT extends AWTAbstractAnalysis {
  public static void main(String[] args) throws Exception {
    AnalysisLauncher.open(new ScatterDemoAWT());
  }

  public void init() {
    Coord3d[] points = new Coord3d[8];

    points[0] = new Coord3d(7.0404835, 905.0, 53.014378);
    points[1] = new Coord3d(7.0043426, 935.0, 53.014378);
    points[2] = new Coord3d(7.0404835, 905.0, 53.014378);
    points[3] = new Coord3d(6.9886136, 885.0, 151.9809);
    points[4] = new Coord3d(7.0043426, 935.0, 53.014378);
    points[5] = new Coord3d(6.96367, 915.0, 153.10077);
    points[6] = new Coord3d(6.9886136, 885.0, 151.9809);
    points[7] = new Coord3d(6.96367, 915.0, 153.10077);

    Scatter scatter = new Scatter(points, Color.BLUE, 5.0f);
    chart = AWTChartFactory.chart(Quality.Advanced);
    chart.getScene().add(scatter);
  }
}
```

the initial picture I've get (without any interaction) with the **current master (a8eab7856bc35b263a9f5c90ae9e4efd4353844e)** is:

![grafik](https://user-images.githubusercontent.com/44067969/113471244-84730f80-945b-11eb-8908-9c04cc423cd5.png)

the initial picture I've get (without any interaction) with the the **commit of this PR(abd99afa6e3992a8d4fc3f06af1d69058771c83f)** is:

![grafik](https://user-images.githubusercontent.com/44067969/113471461-b20c8880-945c-11eb-87e1-39e6e62da726.png)

HannesWell commented:
When looking at my commit again I noticed that the side computation can be fixed with less code changes. Most of the changes come from not recomputing the `unscaledEye` and `unscaledViewPlaneNormal` everytime the side is requested.
Nevertheless I think having a method to `atomically` set the camera position instead of an own methods for every aspect is more intuitive.

jzy3d commented:
Weirdly I can now reproduce the problem you spotted. Beside the left/right jumps, I also note that the text layout is wrong : text should always have a left horizontal alignment when displayed on the right of the chart, and right horizontal alignment when displayed on the left of the chart.

I first finish a feature involving lot of non regression tests to pass to ensure I have a good basis to test your Camera improvements in all dependent jzy3d projects.

Thank you!

HannesWell commented:
That are 'good' news.  At least we can now proceed with a solution. :)
You're right the text alignment is also incorrect.

Alright, just let me know when you are done and I should do anything regarding this issue.
As I said: The API changes and additionally fields for the camera are probably not yet final and need some discussion and the internal computation can also be improved (I just noticed I've forgot a FIXME in the code).

jzy3d commented:
Sorry forgot to answer that : I am indeed in the process of finalizing 2.0 so there is still room for API clarifications and changes, go ahead. As API change is major from 1.* to 2.0, I hope I can include all major API changes now. Note that I wrote [a helper to help users know how to bump to 2.0](https://github.com/jzy3d/jzy3d-api#changes-in-20-version). Your enhancements are welcome there!

For the particular case of your Camera API change I would prefer to not deprecate the independent eye/target/up setter/getter as it is sometime useful to change them independently. I am especially thinking of an upcoming FreeCam navigation mode where I would need to edit camera position parameters separately. Instead of deprecating I would add to the setPosition javadoc indication that these parameters may be edited by {@link #setEye(...)} etc.

Beside that, the addition of the setPosition method is great. I will verify the behavior of the scaling and transforms with the Logarithmic chart integration tests.

HannesWell commented:
I updated the commit as you requested. Two `setPosition()` methods are added and the `setEye()`, `setTarget()` and `setUp()` call `setPosition()` accordingly. Additionally I updated the java-doc accordingly.

In order to minimize the number of required fields now a predicate is used that captures all relevant pre-computed data and allows a quick check without object-creation if a point is on the left side.


In general I have to say that it is not obvious which vector is scaled an which is not. From the code I derived that the `eye` and `target`-vector are scaled while  the `up`-vector is not scaled. I stated my findings in the doc.
But changing this would probably require a greater effort.

jzy3d commented:
Hi! Good news I confirm it is working well on all my projects, thank you!

The only bad thing is that I haven't handled the PR properly seamingly. I fetch locally, edit few things. My push did not go to your branch so I may redo the process.

You might be interested to review the [unit test I added](https://github.com/jzy3d/jzy3d-api/commit/bca46f7dad19608ab053652fd81ef08af6574e14#diff-1dcf5ee8ed105626c6b2f3c19a132444d981a35a62152e9a6344bd8a3b21e6ee). Probably much better than doing an ITTest.

jzy3d commented on jzy3d-core/src/main/java/org/jzy3d/plot3d/rendering/view/Camera.java:
```diff
@@ -182,10 +248,33 @@ public float getFar() {
+      // Not scaled but normalized to length one for numerical stability.
+      Coord3d vpn = up.getNormalizedTo(1f).cross(viewDirectionU); // view-plane normal
+
+      isOnLeftSide = p -> 0 < (p.x - e.x) * vpn.x + (p.y - e.y) * vpn.y + (p.z - e.z) * vpn.z;
```
I just did not get why you create the function on the fly and record it as a field? I would find more readable to directly execute the code. The fact that the function is cached let me think that lines 267 to 272 will be skipped for the second call to side() hence leading to an old eye/target/scale value.

My point is mainly on the code readability as I tried your code and it really works better than before.

HannesWell replied:
The purpose of the Predicate-lambda is to capture the Coord3d objects `e` and `vpn` (as de-facto fields of the Predicate object), in order to avoid re-creating them every time side is called. These Coord3d-objects/vectors do not change while eye/target/up/scale does not change. Since they are captured and the predicate object is re-used (as you already noticed)  runtime is saved since their values are not re-computed and the garbage-collector does not have to clean up a few Coord3d objects for each call to `side()`. So yes, you're right that line 267 to 272 are skipped intentionally when `isOnLeftSide` is `null`.

But `isOnLeftSide` is set to `null` every time `setPosition()` is called (see line 204) in order to trigger a re-computation of the `isOnLeftSide` `Predicate` object. And because each of the fields `eye`, `target`, `up` and `scale` is only set through setPosition() the `isOnLeftSide` never operates on an old value of `eye`, `target`, `up` and `scale`.

In my first commit I created two fields for the values of `e` and `vpn`. But with the intention to have all calculations at one place closely together and to minimize the number of fields I decide to use this approach and tried to make it clear with the comment at the beginning of the if-block.

So if you find this has poor readability I can change this (I think readability is very important). I just saw that the variable `viewDirectionU` is a bit off the schema of the other variables so I would like to fix this anyway.

jzy3d replied:
Thanks. Just extended the unit test to reflect this.

I forgot to answer you that you are right about scaling the up vector. Not scaling it is not consistent. However it works because the up vector is simply a vertical direction - changing its length does not change anything for the underlying opengl primitive.

jzy3d commented:
Thank you for the fix.
I added a unit test to Camera.side() [here](https://github.com/jzy3d/jzy3d-api/commit/ac91bd8fdd8e9d50cd0a5e41de3a7628ca4bd9df). Note that it is really naïve and it was not failing before I merge your code (so it does not mimic your example).

HannesWell commented:
Thanks for reviewing and that it made it eventually into the master. 👍🏽
